actors/evm/src/interpreter/precompiles/bls_util.rs

```diff
@@ -0,0 +1,421 @@
+    // SAFETY: both inputs are valid blst types
+    unsafe { blst_p1_to_affine(&mut p_affine, p) };
+    p_affine
+}
+
+#[inline]
+pub fn p1_from_affine(p_affine: &blst_p1_affine) -> blst_p1 {
+    let mut p = blst_p1::default();
+    // SAFETY: both inputs are valid blst types
+    unsafe { blst_p1_from_affine(&mut p, p_affine) };
+    p
+}
+
+#[inline]
+pub fn p2_to_affine(p: &blst_p2) -> blst_p2_affine {
+    let mut p_affine = blst_p2_affine::default();
+    // SAFETY: both inputs are valid blst types
+    unsafe { blst_p2_to_affine(&mut p_affine, p) };
+    p_affine
+}
+
+#[inline]
+pub fn p2_from_affine(p_affine: &blst_p2_affine) -> blst_p2 {
+    let mut p = blst_p2::default();
+    // SAFETY: both inputs are valid blst types
+    unsafe { blst_p2_from_affine(&mut p, p_affine) };
+    p
+}
+
+pub fn read_scalar(input: &[u8]) -> Result<blst_scalar, PrecompileError> {
+    if input.len() != SCALAR_LENGTH {
+        return Err(PrecompileError::IncorrectInputSize);
+    }
+
+    let mut out = blst_scalar::default();
+    // SAFETY: `input` length is checked previously, out is a blst value.
+    unsafe {
+        // Note: We do not use `blst_scalar_fr_check` here because, from EIP-2537:
+        //
+        // * The corresponding integer is not required to be less than or equal than main subgroup
+        // order `q`.
+        blst_scalar_from_bendian(&mut out, input.as_ptr())
+    };
+
+    Ok(out)
+}
+
+/// Performs a G1 scalar multiplication
+///
+/// Takes a G1 point in affine form and a scalar, and returns the result
+/// of the scalar multiplication in affine form
+///
+/// Note: The scalar is expected to be in Big Endian format.
+#[inline]
+pub fn p1_scalar_mul(p: &blst_p1_affine, scalar: &blst_scalar) -> blst_p1_affine {
+    // Convert point to Jacobian coordinates
+    let p_jacobian = p1_from_affine(p);
+
+    let mut result = blst_p1::default();
+
+    // SAFETY: all inputs are valid blst types
+    unsafe { blst_p1_mult(&mut result, &p_jacobian, scalar.b.as_ptr(), scalar.b.len() * 8) };
+
+    // Convert result back to affine coordinates
+    p1_to_affine(&result)
+}
+
+/// Performs a G2 scalar multiplication
+///
+/// Takes a G2 point in affine form and a scalar, and returns the result
+/// of the scalar multiplication in affine form
+///
+/// Note: The scalar is expected to be in Big Endian format.
+#[inline]
+pub fn p2_scalar_mul(p: &blst_p2_affine, scalar: &blst_scalar) -> blst_p2_affine {
+    // Convert point to Jacobian coordinates
+    let p_jacobian = p2_from_affine(p);
+
+    let mut result = blst_p2::default();
+    // SAFETY: all inputs are valid blst types
+    unsafe { blst_p2_mult(&mut result, &p_jacobian, scalar.b.as_ptr(), scalar.b.len() * 8) };
+
+    // Convert result back to affine coordinates
+    p2_to_affine(&result)
+}
+
+/// Encodes a G2 point in affine format into byte slice with padded elements.
+/// G2 points have two coordinates (x,y) where each coordinate is a complex number (real,imaginary)
```

Comment:
```suggestion
/// G2 points have two coordinates (x,y).  Each coordinate is in field Fp2.  Each element of Fp2 is made up of two elements of Fp.  Therefore a useful analogy is that each coordinate is made up of a complex number (real,imaginary)
```

Reply:
Apparently there are some good analogies to real and complex in the Fp2 field extension but I'd rather we clarify that this is not exactly true